### DISCUSSION TITLE
IBKR TWS API 5 Second delay to collect bar data upon close

### DISCUSSION THREAD

**Discussion by surfandkite** [ib-api-reloaded/ib_async#87] [Q&A]
Hello, does anyone else encounter the issue of a 5 second delay for all data requests from IBKR TWS API? Using the Python script below which just prints 1min bar data upon close for /nq 1min bars I get a 5 second delay while it is near instant for my developer. Attached is a screenshot of my terminal. Notice all results take 5+ seconds. 
![Screenshot 2024-10-31 130404](https://github.com/user-attachments/assets/80e71c2d-cb2e-4083-b700-92843b6e84b0)

```python
from ib_insync import *
import datetime as dt


def on_bar(bars: BarDataList, has_new_bar: bool):
    if has_new_bar:
        closed_bar = bars[-2]
        print(f"{dt.datetime.now()}: {closed_bar}")


ib = IB()
ib.connect("localhost", 7497, 2)

contract = Future("NQ", "202412", "CME", currency="USD")
ib.qualifyContracts(contract)

b = ib.reqHistoricalData(contract, "", "300 S", "1 min", "TRADES", False, 1, True)
b.updateEvent += on_bar

while True:
    ib.sleep(1)
```

**Comment by mattsta** [comment]
It looks like other people have noticed the same problem here: https://www.elitetrader.com/et/threads/ib-real-time-data-lag.343165/

Maybe that's just how it works? Some of the IBKR data APIs work only on 5 second time increments, so maybe their network sender bundles all previously closed bars on their next 5 second timer.

For live data, I use `reqRealTimeBars()` but I never tracked the _exact_ arrival time of the data versus the timestamps being returned (any processing needing less than 15 second latency is better served through streaming prices and not bars anyway).

The fastest price data is always from `reqMktData()` for any live bid/ask/last/size processing though. It updates as fast as 250 ms (IBKR drops any updates faster than 250 ms) but price data updating 4 times per second is faster than you can practically use the IBKR order APIs anyway over a standard public network.

**Comment by Jeffreylabranche** [comment]
did you found a solution ? I have the same 5 seconds delay for 1min and 15 min bars. I'm wondering if it is related to more recent IB Gateway settings since I'm running my code on a Linux VM (Ubuntu 24.04 LTS) with older IB Gatway version (10.19) without delay but if II run the code on a most recent IB Gateway version (10.30) I get the 5 seconds delay.

**Reply by mattsta** [reply]
This is just a limit of IBKR itself. The IBKR API only creates a new subscribed bar every 5 seconds and only sends a new market data update every 250 ms.

For faster updates, you would have to subscribe to a real market data service like polygon.io or consuming OPRA/CME/etc feeds directly.

On the other side, the delays are actually okay because there is almost nothing useful you can do on such short time frames because of the general noise present in second-by-second trade matching.

**Reply by Jeffreylabranche** [reply]
Thank you @mattsta for the quick answer. Makes sense and I agree with you that over such a short time period the 5 seconds delay is not a big deal considering all the noise. Although I'm still curious to see if someone else in the community noticed this delay with most recent IB Gateway versions because with version 10.19 I was able to receive bars in real time (current time = bar timestamp). I won't pursue the investigation on my side but might post a code sample tonight if it can help others to investigate the difference.

**Comment by surfandkite** [comment]
Never found a solution unfortunately. Prior to July 2024 I didn’t the 5 second delay. Tried to run it on older versions of IBKR’s static TWS but still had the issue. Sent from my iPhoneOn Jan 5, 2025, at 9:11 AM, Jeffrey Labranche ***@***.***> wrote:﻿
Thank you @mattsta for the quick answer. Makes sense and I agree with you that over such a short time period the 5 seconds delay is not a big deal considering all the noise. Although I'm still curious to see if someone else in the community noticed this delay with most recent IB Gateway versions because with version 10.19 I was able to receive bars in real time (current time = bar timestamp). I won't pursue the investigation on my side but might post a code sample tonight if it can help others to investigate the difference.

**Reply by Jeffreylabranche** [reply]
Here is the code I have used and below I share the results. We can see that with IB gateway 10.19 I'm able to get 1 min bars without delay.



```
from ib_async import *
import nest_asyncio
import datetime
import os
import sys
import pytz
nest_asyncio.apply()

script_name=os.path.basename(__file__)

ib = IB()
ib.connect('127.0.0.1', 4001, clientId=1) 

ticker='USDCAD'
contract = Forex(ticker, 'IDEALPRO')
ib.qualifyContracts(contract)

def on_new_bar_1min(bars, has_new_bar):
    if has_new_bar:
        latest_bar_1min = bars[-1]
        current_time_1min = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        print(f"Current time: {current_time_1min}, 1 min bar timestamp: {latest_bar_1min.date}, Close: {latest_bar_1min.close}")


def on_new_bar_5sec(bars, has_new_bar):
    if has_new_bar:
        latest_bar = bars[-1]
        eastern = pytz.timezone('US/Eastern')
        bar_time_eastern = latest_bar.time.astimezone(eastern)
        current_time = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        print(f"Current time: {current_time}, 5 seconds bar timestamp: {bar_time_eastern.strftime('%Y-%m-%d %H:%M:%S')}, Close: {latest_bar.close}")

def subscribe_1min():
    contract = Forex(ticker)
    data_1min = ib.reqHistoricalData(
        contract,
        endDateTime='',
        durationStr='1 D',
        barSizeSetting='1 min',
        whatToShow='MIDPOINT',
        useRTH=False,
        keepUpToDate=True,
        timeout=0
    )
    data_1min.updateEvent.clear()
    data_1min.updateEvent += on_new_bar_1min
    print('Subscribed 1min')


data_5sec = ib.reqRealTimeBars(contract, 5, 'MIDPOINT', False)
data_5sec.updateEvent.clear()
data_5sec.updateEvent+=on_new_bar_5sec

subscribe_1min()


ib.run()
```
**IB Gateway 10.19 results:**
<img width="697" alt="Screenshot 2025-01-05 at 8 18 08 PM" src="https://github.com/user-attachments/assets/258b8ae6-291e-41fc-ada4-560f564afc25" />

IB Gateway 10.30 results
<img width="798" alt="Screenshot 2025-01-05 at 8 19 06 PM" src="https://github.com/user-attachments/assets/f04d5d77-2536-4e40-9444-cb6c9590b1cc" />

**Reply by alittlenova** [reply]
I met the same problem but I cannot find any place to download the old version of gateway now. Do you have the installer for windows or linux? May I have your old version of installer?

**Reply by alittlenova** [reply]
I did several day's deep investigation on this problem because the 5 seconds delay is realy a big problem for my trading algorism. I cannot accept even 1 second delay if you claim it to be a 'real time' data. Lucky I finally figured it out with my v1037 gateway. One need to close the bar with the timestamp xx:xx:55 which means the timestamp of the bar is actually the start point of the statistic period. If you start to accumulate a real time bar from timestamp=xx:xx:00 and finish the statistics on the last bar with timestamp=xx:xx:55, then the total volume you obtained on these 12 bars will be equavalent to the Historical data's result on xx:xx:00 minute bar. So it is not because of the 5 seconds delay, it is just because the timestamp is 5 seconds ago. Each time bar records the time between the -5s to 0s, and marked timestamp on -5s.

**Reply by Jeffreylabranche** [reply]
Thanks a lot  @alittlenova  for sharing your findings! I guess that means the below method needs to me modified to take into consideration the 5 seconds offset ? Would you mind sharing how you adjusted your code ?

`ib.reqHistoricalData(
        contract,
        endDateTime="",
        durationStr=duration,
        barSizeSetting=timeframe,
        whatToShow="TRADES",
        useRTH=True,
        keepUpToDate=False,
        timeout=0,
    )`

**Reply by alittlenova** [reply]
You dont have to use reqHistoricalData. If you want to use 5s real-time bar data, just notice that a line with timestemp hh:mm:ss means the trades between hh:mm:ss and hh:mm:(ss+5) has been take into statistics. For example, when you received a 5s bar with timestamp 15:02:15, it means the statistic period begin at 15:02:15 and end at 15:02:20. If you received this data on 15:02:20, that means you almost immediately knows what happens before 15:02:20. You cannot receive a 5s bar before the bar ending.
Historical data is the same, they are constructed by real-time data on the server, and always update later than real-time data.

**Reply by alittlenova** [reply]
If you want to verify this, you need reqHistoricalData. use it to obtain the 1 min bar for some contract. You will then have some bars with timestamp and price and volume. You need to focus on timestamp and volume. These numbers will never change on IBKR server. Now for example you have received some 5s real-time data which covers one minute that you can obtain with reqHistoricalData also. You assemble the 5s real-time data from hh:mm:00 to hh:mm:55, their volume summary would be exactly equivalent to 1 min historical bar with timestamp hh:mm:00.

**Reply by Jeffreylabranche** [reply]
Thank you for the precision @alittlenova. That means if I want to use 1min bar or any other timeframe I need to buid the method myself and cannot use reqHistoricalData with flag KeepUpToDate=True unless the source code is modified.
